### DISCUSSION TITLE
Streaming custom LLM with NemoGuardrails

### DISCUSSION THREAD

**Discussion by nicolemzh** [NVIDIA-NeMo/Guardrails#432] [Q&A]
What are the requirements to stream a custom LLM with NemoGuardrails?

The custom LLM is working with NemoGuardrails without the streaming function. However, I'm having an issue implementing streaming with stream_async or the StreamingHandler. I was wondering if anyone has tried this task and if there are extra steps to stream with a custom LLM compared to GPT or LangChain through NemoGuardrails.

**Comment by drazvan** [comment]
Hi @nicolemzh ! The NeMo LLM provider is a good example that also supports streaming: https://github.com/NVIDIA/NeMo-Guardrails/blob/develop/nemoguardrails/llm/providers/nemollm.py. In a custom LLM, it needs to be implemented similarly. Let me know if you need more guidance.

**Reply by nicolemzh** [reply]
Thanks @drazvan, appreciate your help! To give more context, we want to implement streaming with mixtral-8x7b-32768 through the Groq provider. Since we tried using Groq using Langchain (https://python.langchain.com/docs/integrations/chat/groq/) but it is not a listed provider, we decided to register ChatGroq as a "custom LLM" with the same methods, and also taking in a BaseChatModel (https://github.com/langchain-ai/langchain/blob/3856dedff469b2ce08e3d416e9e351e2b8950e27/libs/partners/groq/langchain_groq/chat_models.py):

1. We added "streaming: True" and specified our model in the .yml file
2. We registered ChatGroq as a custom LLM
3. implemented streaming the same way as before (was working with GPT)

However, when I send a request using the streaming implementation, I get a response, but it is not streaming and is returned as the whole message.

I was wondering why this is because we were running into the same issue when streaming GPT with NemoGuardrails in the past. The request would stream perfectly fine using gpt-3.5-turbo-instruct but would not stream when we changed the model to gpt-4 in the .yml file.

**Reply by drazvan** [reply]
@nicolemzh : Sorry for the delayed follow-up on this. Can you double check that the streaming flag is set correctly here: https://github.com/NVIDIA/NeMo-Guardrails/blob/develop/nemoguardrails/rails/llm/llmrails.py#L358. (add a breakpoint/print statement there). 

Also, is the behavior with streaming with gpt-3.5-turbo-instruct vs. gpt-4 still reproducible? It should work with both.

**Reply by JanviAg** [reply]
@nicolemzh  I am also building custom LLM using Nemo but I am unable to run the code. Can you help me out.

**Reply by trebedea** [reply]
Hi @JanviAg , 
Are you using an LLM not supported by Langchain?

**Reply by JanviAg** [reply]
@trebedea Yes I am using gemini LLM, which is not supported by Langchain

**Reply by trebedea** [reply]
@JanviAg , Langchain has an integration with Google models, including Gemini:  https://python.langchain.com/docs/integrations/chat/google_generative_ai/

It is way easier to use this wrapper than create one yourself. We can assist you if you really need to create a new one (with minimal functionality).

**Reply by JanviAg** [reply]
@trebedea  I am also using aws. I already have completion end points implemented in my project. By using custom llm I wanted to use my completions. This is possible by using custom llm, is my understanding correct?

**Reply by trebedea** [reply]
I need more details to understand the LLM providers you are using. Creating custom LLM interfaces for NeMo Guardrails should only be needed when there is no existing Langchain provider. Amazon Bedrock already has a Langchain provider: https://python.langchain.com/docs/integrations/chat/bedrock/

Even the Nemo LLM example for NVIDIA models has been removed as there is a Langchain integration available now.

**Reply by JanviAg** [reply]
@trebedea

That makes sense. In my case, I’m not using the native Google or Bedrock SDKs—instead, I have custom completion endpoints for both Gemini and AWS models, exposed via my own LLM Gateway service.

These endpoints follow an OpenAI-compatible completion format, so my goal is to let NeMo Guardrails call them directly through a custom LLM provider.

I understand that if I were using the official Google or Bedrock APIs, I could simply rely on LangChain’s existing integrations (like ChatGoogleGenerativeAI or ChatBedrock).

However, since these endpoints are internal and not part of LangChain’s supported list, could you please confirm if creating a minimal CustomLLM subclass (inheriting from BaseLLM) is still the recommended approach for integrating such custom endpoints with NeMo Guardrails?

**Reply by trebedea** [reply]
If your endpoints are OpenAI compatible, the simplest thing is to use the OpenAI Langchain wrapper and just provide your specific endpoint URL as a parameter for `base_url` attribute. The docs are not explicit, but this is the intent of `base_url`: https://python.langchain.com/api_reference/openai/llms/langchain_openai.llms.base.OpenAI.html
